runtime/sema/access.go

```diff
@@ -0,0 +1,190 @@
+/*
+ * Cadence - The resource-oriented smart contract programming language
+ *
+ * Copyright Dapper Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package sema
+
+import (
+	"github.com/onflow/cadence/runtime/ast"
+	"github.com/onflow/cadence/runtime/common/orderedmap"
+)
+
+type Access interface {
+	isAccess()
+	// returns whether receiver access is less permissive than argument access
+	IsLessPermissiveThan(Access) bool
+	// returns whether receiver access permits argument access
+	PermitsAccess(Access) bool
+	Access() ast.Access
+}
+
+type EntitlementSetKind uint8
+
+const (
+	Conjunction EntitlementSetKind = iota
+	Disjunction
+)
+
+type EntitlementAccess struct {
+	astAccess    ast.EntitlementAccess
+	Entitlements *EntitlementOrderedSet
+	SetKind      EntitlementSetKind
+}
+
+var _ Access = EntitlementAccess{}
+
+func NewEntitlementAccess(
+	astAccess ast.EntitlementAccess,
+	entitlements []*EntitlementType,
+	setKind EntitlementSetKind,
+) EntitlementAccess {
+	set := orderedmap.New[EntitlementOrderedSet](len(entitlements))
+	for _, entitlement := range entitlements {
+		set.Set(entitlement, struct{}{})
+	}
+	return EntitlementAccess{
+		Entitlements: set,
+		SetKind:      setKind,
+		astAccess:    astAccess,
+	}
+}
+
+func (EntitlementAccess) isAccess() {}
+
+func (a EntitlementAccess) Access() ast.Access {
+	return a.astAccess
+}
+
+func (e EntitlementAccess) PermitsAccess(other Access) bool {
```

Comment:
This function essentially implements the core logic for comparing entitlements:
* If an entitlement set `X` permits access of an entitlement set `Y`, that effectively means that a reference authorized to `Y` is a subtype of a reference authorized to `X`. 
* Equivalently, `X` permitting `Y` means that a `Y`-entitled reference is permitted to access an `X`-access member on a composite. 
* Equivalently, an access modifier of `X` on a composite is a valid implementation of the same member declared with `Y` access on an interface (access is contravariant). 

There are 4 essential combinations of checks we might need to perform when determining whether one set permits another depending on whether the two sets are conjunctions, disjunctions, or one of each.